[#IOPID-3131] Common monorepo package, include types and repo to interact with io-auth-sessions-topic ServiceBus topic

giamma1295 opened the pull request:
<!--- Please always add a PR description as if nobody knows anything about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the same page. -->
<!--- Write this PR as you were contributing to a public OSS project, where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->
Shared modules, helper functions, and types for authentication and identity, used across multiple applications within the monorepo:
- Shared `io-ts` types.
- Reusable helpers.
- Common Repositories (eg. a Repo wich allow write events in a ServiceBus Topic).
- and other commonly used components.

Proposed common package structure:

```
├── README.md
├── dist // build output
├── package.json
├── src
│   ├── repositories // Here will be the various repos commonly used in the monorepo
│   |   ├── __tests__
│   |   │   └──<repo_name>.test.ts // repository unit tests
│   │   └── <repo_name>.ts // repository implementation
│   ├── types // Here will be the various types commonly used in the monorepo
│   |   ├── __tests__
│   |   │   └── <type_name>.test.ts // type unit tests(eg. decoding/encoding tests)
│   |   └── <type_name>.ts // type definition
│   └──<components_type> // In future other components type can be placed
├── sonar-project.properties
├── tsconfig.eslint.json
├── tsconfig.json
└── vite.config.mts

```

This first release includes:
- `AuthSessionEvent` -> an `io-ts` type for validating and mapping ServiceBus `io-auth-sessions-topic` messages.
- `AuthSessionsTopicRepository` -> a repository for interacting with `io-auth-sessions-topic`.

Those common components will be used in:
- `io-profile-async` -> in a new ServiceBusTrigger Azure Function, to validate and map retrieved events from topic.
- `io-session-manager` -> in order to emit new Authentication Events on Log-In and Log-Out occurrences.
- `io-session-manager-internal` -> in order to emit new Authentication Events on Log-Out occurrences.

#### List of Changes
- Init `@pagopa/io-auth-n-identity-commons` package.
- Configure SonarCloud for the newly created package.
- Add `AuthSessionEvent` and related `io-ts` types.
- Add `AuthSessionsTopicRepository`

<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Proposal for a new package to be used as an internal dependency to promote reuse and consistency across monorepo projects.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

#### Screenshots (if appropriate):

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a changeset, which I've added.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

giamma1295 committed:
bootstrap team common package

giamma1295 committed:
update yarn.lock

giamma1295 committed:
add README

giamma1295 committed:
add README

giamma1295 committed:
settingUp exports

giamma1295 committed:
cleanUp

giamma1295 committed:
add auth sessions ServiceBus topic repository

giamma1295 committed:
lint cleanUp

giamma1295 committed:
add vitest-configuration

giamma1295 committed:
setup sonarcloud

giamma1295 committed:
clean tsconfig

giamma1295 committed:
[#IOPID-3132] Update pre_commit_script (#329)

<!--- Please always add a PR description as if nobody knows anything
about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the
same page. -->
<!--- Write this PR as you were contributing to a public OSS project,
where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->
Update `pre_commit_script` from  version `0.0.1` to `0.1.0`
#### List of Changes
<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how
-->
<!--- your change affects other areas of the code, etc. -->

#### Screenshots (if appropriate):

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all
the boxes that apply: -->
- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes
that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're
here to help! -->
- [ ] My change requires a changeset, which I've added.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

giamma1295 committed:
sonar code-review ci integration for io-auth-n-identity-commons package

giamma1295 committed:
add repository tests

giamma1295 committed:
test cleanUp

giamma1295 committed:
add repo test

giamma1295 committed:
provisional changeset

giamma1295 committed:
Merge branch 'main' into IOPID-3131

giamma1295 committed:
cleanUp

giamma1295 committed:
clean package.json from unused dependencies

giamma1295 committed:
fix typos

giamma1295 committed:
test cleanup

giamma1295 committed:
test cleanup2

giamma1295 committed:
cleanUp 3

giamma1295 committed:
remove not actually present module prefixes from exclusions

salvatorelaiso committed:
[#IOPID-3125] Sonar coverage exclusions configuration (#328)

<!--- Please always add a PR description as if nobody knows anything
about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the
same page. -->
<!--- Write this PR as you were contributing to a public OSS project,
where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes
<!--- Describe your changes in detail -->
Update Sonar code coverage exclusions for
- `io-fast-login` 40e518d54f542a4a0e9e4324d2ec3ccff3e7d019
- `io-lollipop` 0a13a75daea0b71eb1d2ae9b0dc2d963431ab64f
- `io-profile` 5bd94126aaf20c57e53e8deea55508d77a9b7e14
- `io-public` daaf49f9a99703e8a14dad529c3261107cf6469d
- `io-session-manager-internal` 28594284b35c31c36494aab8d68dda0ac2618824
- `io-web-profile` 4a6a11c63ba57ae462bd8d0587e965ae664d4acd

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how
-->
<!--- your change affects other areas of the code, etc. -->

#### Screenshots (if appropriate):

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all
the boxes that apply: -->
- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes
that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're
here to help! -->
- [ ] My change requires a changeset, which I've added.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

arcogabbo reviewed (commented):
LGTM, minor comments

gquadrati commented on packages/io-auth-n-identity-commons/src/repositories/auth-sessions-topic-repository.ts:
```diff
@@ -0,0 +1,30 @@
+export type AuthSessionsTopicRepositoryDeps = {
+  authSessionsTopicSender: ServiceBusSender;
+};
+const emitSessionEvent: (
```
Can we try to make it generic?

giamma1295 replied:
can we do this in an another PR?

giamma1295 committed:
move .env.common -> .env.common.example, ignoring .env.common file (#331)

<!--- Please always add a PR description as if nobody knows anything
about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the
same page. -->
<!--- Write this PR as you were contributing to a public OSS project,
where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes
<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how
-->
<!--- your change affects other areas of the code, etc. -->

#### Screenshots (if appropriate):

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all
the boxes that apply: -->
- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes
that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're
here to help! -->
- [ ] My change requires a changeset, which I've added.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

gquadrati reviewed (approved):
LGTM. I've Just left a small comment, totally optional — up to you.

giamma1295 committed:
Merge branch 'main' into IOPID-3131

giamma1295 commented on yarn.lock:
```diff
@@ -7045,6 +7212,13 @@ __metadata:
   languageName: node
   linkType: hard
 
+"fp-ts@npm:^2.16.10":
```
There is a bit too much entropy in the monorepo regarding the version of fp-ts, it should be normalized, however the highest used is v `2.16.9`, the only difference between this and the new point release, `2.16.10`, is `Add apFirstW and apSecondW to ReaderTask, #1958`, it should be safe to use it.
In any case I downgrade to 2.16.9

giamma1295 committed:
cleanUp

giamma1295 committed:
fix package.json

giamma1295 committed:
update yarn lock

giamma1295 commented on packages/io-auth-n-identity-commons/package.json:
```diff
@@ -0,0 +1,41 @@
+    "build": "tsc",
+    "build:watch": "tsc -w",
+    "typecheck": "tsc --noemit",
+    "import": "node ./dist/app.js",
```
leftover, removed
